test/integration/provider/alicloud.go

```diff
@@ -0,0 +1,190 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package provider
+
+import (
+	"context"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	api "github.com/gardener/machine-controller-manager-provider-alicloud/pkg/alicloud/apis"
+	"github.com/gardener/machine-controller-manager-provider-alicloud/pkg/spi"
+	"github.com/gardener/machine-controller-manager/pkg/apis/machine/v1alpha1"
+	"github.com/gardener/machine-controller-manager/pkg/util/provider/driver"
+
+	providerDriver "github.com/gardener/machine-controller-manager-provider-alicloud/pkg/alicloud"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/json"
+	"log"
+)
+
+/**
+	Orphaned Resources
+	- VMs:
+		Describe instances with specified tag name:<cluster-name>
+		Report/Print out instances found
+		Describe volumes attached to the instance (using instance id)
+		Report/Print out volumes found
+		Delete attached volumes found
+		Terminate instances found
+	- Disks:
+		Describe volumes with tag status:available
+		Report/Print out volumes found
+		Delete identified volumes
+**/
+
+func newSession(machineClass *v1alpha1.MachineClass, secret *v1.Secret) spi.ECSClient {
+	var (
+		providerSpec *api.ProviderSpec
+		sPI          spi.PluginSPIImpl
+	)
+
+	err := json.Unmarshal([]byte(machineClass.ProviderSpec.Raw), &providerSpec)
+	if err != nil {
+		providerSpec = nil
+		log.Printf("Error occured while performing unmarshal %s", err.Error())
+	}
+	sess, err := sPI.NewECSClient(secret, providerSpec.Region)
+	if err != nil {
+		log.Printf("Error occured while creating new session %s", err)
+	}
+	return sess
+}
+
+func getMachines(machineClass *v1alpha1.MachineClass, secretData map[string][]byte) ([]string, error) {
+	var machines []string
+	var sPI spi.PluginSPIImpl
+	driverProvider := providerDriver.NewAlicloudPlugin(&sPI)
+	machineList, err := driverProvider.ListMachines(context.TODO(), &driver.ListMachinesRequest{
+		MachineClass: machineClass,
+		Secret:       &v1.Secret{Data: secretData},
+	})
+	if err != nil {
+		return nil, err
+	} else if len(machineList.MachineList) != 0 {
+		for _, machine := range machineList.MachineList {
+			machines = append(machines, machine)
+		}
+	}
+	return machines, nil
+}
+
+// getOrphanedInstances returns list of Orphan resources that couldn't be deleted.
+func getOrphanedInstances(tagName string, tagValue string, machineClass *v1alpha1.MachineClass, secretData map[string][]byte) ([]string, error) {
+	sess := newSession(machineClass, &v1.Secret{Data: secretData})
+	var instancesID []string
+	var tags = &[]ecs.DescribeInstancesTag{{Key: tagName, Value: tagValue}}
+	input := ecs.CreateDescribeInstancesRequest()
+	input.InstanceName = "instance-state-name"
+	input.Status = "running"
+	input.Tag = tags
+
```

Comment:
When and where these orphan resources are created? 
why instance name is hardcoded.

Reply:
These orphan resources are/might get created during the Integration test. Also, the instance name was there when I tested something and forgot to remove it.
Thanks, I will be removing this, as it is not required.